src/py2spack/conversion_tools.py

```diff
@@ -64,19 +62,7 @@ def requirement(self) -> str | None:
 
 
 # TODO @davhofer: integrate this into a datastructure, omit mutable global state  # noqa: TD003
-evalled: dict = {}
-
-
-def acceptable_version(version: str) -> pv.Version | None:
-    """Check whether version string can be parsed and has correct format."""
-    try:
-        v = pv.parse(version)
-        # do not support post releases of prereleases etc.
-        if v.pre and (v.post or v.dev or v.local):
-            return None
-        return v
-    except pv.InvalidVersion:
-        return None
+evalled: dict[tuple[str, specifiers.SpecifierSet], sv.VersionList] = {}
```

Comment:
Global state here, this code is taken from https://github.com/spack/pypi-to-spack-package/.
Should probably try to refactor this to remove mutable global state.

Reply:
The function `_pkg_specifier_set_to_version_list` that uses this dict takes an additional argument, a PyProjectProvider object,which is used to look up the package information. By default this option is not hashable. But given a specific PyProjectProvider, e.g. PyPI, or Github, the output of the function should always be the same. So maybe I could add a custom `__hash__` function to the class which returns some PyProjectProvider instantiation identifier string (e.g. "github" or "pypi" or "test.pypi", or the base url).

Reply:
Solved using dataclasses